### ISSUE TITLE
Update query string without state reload

### ISSUE THREAD

**Issue by SinisaG** [angular-ui/ui-router#3280]
Hey guys, 

thanks for all the hard work on this library. I have a question, regarding modifying url (path or query string), without modifying/reloading the state. 

**Use case:** We go to specific URL and use `resolve` to fetch some data. Based on that data, we append cache breaking/or some other params to our url as query string. Sometimes this happens inside of controller (if change is caused by specific event). 

So basically our use case is modifying query string inside of `resolve` or `controller`. 

**What we want:**   The desired outcome would be to do this, without reloading the state. 

**What I've tried:** 

1.)  In `resolve` or `controller` - Setting `reloadOnSearch` to `false` and update query string with `$location.search().`

```
     $location.search('period', params.period);
     $location.search('variable', params.variable);
```

2.) In `resolve` - Use `$state.go` with updated params object and `notify:false` option.

```
      $state.go('cmsList.transfers.competition', params, notify: false);
      d.reject({
           message: 'We meed to update the url'
       });
```

3.) In service, once we have everything resolved - Use `$state.transitionTo`, with updated params object and special options. 
```
      $state.transitionTo($state.current.name, params, {locations: 'replace', reload: false, notify: false});
```

All the above behave in the same way, they start a new `state` transition and cancel the current state (if I do something like this inside of state `resolve`) or they reload the state they are in (if I do it after I resolve all data - in controller or service). That is bad, because it fetches all the needed data again in the resolve and that feels very wasteful.

Maybe more general question would be, how to properly do redirects inside of `resolve` and how to update query string based on user input, that would not cause any reload? Maybe some of this behavior is expected generally, but I just don't know it. :) 

Thank you for your help! I am happy to clarify this further if needed, of course. 

Stack: 
* AngularJS - 1.6.1
* ui-router - 1.0.0-rc.1

**Comment by SinisaG** [comment]
I've created this Plunker to try and replicate the issue: http://embed.plnkr.co/ZPyn8XobLeYs9IvZSRz9/

But it turns out parent resolve is not executed twice in this case. I see the state change though (through visualizer - it actually does not change the state), when changing the query string, but I guess that is normal?

**Comment by SinisaG** [comment]
I ended defining `redirectTo` function on `state` declaration. There I request my resolve to be executed and when promise is resolved, I redirect to the same state, with updated params (under some condition, not to cause a loop). 

Code: 
```
redirectTo: function (transition) {
   //check url
    if (transition.params().variable === undefined) {
        //get resolve
        var resolvePromise = transition.injector().getAsync('initialData');
       // get the service I need 
        var MatchUrlRewriteService = transition.injector().get('MatchUrlRewriteService');
        return resolvePromise.then(function (initialData) {
            //redirect --> This function returns object {state: 'name', params: params}
            return MatchUrlRewriteService.resolveMatchUrl(initialData.matchFeed.match);
          });
      }
 }
```


That works fine and resolve is executed only once, which is a major benefit, because we do some expensive operations there. But this still throws an error in console (visualizer shows state redirect): 

```
Possibly unhandled rejection: {"type":2,"message":"The transition has been superseded by a different transition" ....
```

I think this should not throw an error, but maybe I am missing something? 

Anyway, my initial problem was solved, but if anyone encounters the same issue, would be nice to hear alternative approach to that. And if creators of library can share your point of view, that would be interesting to hear as well.

**Comment by christopherthielen** [comment]
There's a lot of things to respond to here... I'll do my best:

> Use case: We go to specific URL and use resolve to fetch some data. Based on that data, we append cache breaking/or some other params to our url as query string. 
> So basically our use case is modifying query string inside of resolve or controller.
> What we want: The desired outcome would be to do this, without reloading the state.

[Use a `dynamic` parameter](https://ui-router.github.io/ng1/docs/latest/interfaces/params.paramtypedefinition.html#dynamic).  Define a query parameter and declare it as `dynamic`.  Transition to the same state but change the param value.

```js
.state('foo', {
  url: '/foo?cachebust', 
  params: { cachebust: { dynamic: true } }
});
```

Now, changes to `cachebust` parameter won't cause `foo` to reload.  You can do this in the controller or in a transition hook like  `onEnter`.  For example: 

```js
.state('foo', {
  url: '/foo?cachebust', 
  params: { cachebust: { dynamic: true } },

  onEnter: function($transition$, $state) {
    $transition$.promise.then(function() {
      $state.go('foo', { cachebust: Math.random() });
    });
  }
});
```


> All the above behave in the same way, they start a new state transition and cancel the current state (if I do something like this inside of state resolve) or they reload the state they are in (if I do it after I resolve all data - in controller or service). That is bad, because it fetches all the needed data again in the resolve and that feels very wasteful.

> Maybe more general question would be, how to properly do redirects inside of resolve and how to update query string based on user input, that would not cause any reload?

Great question, you should [use `redirectTo`](https://ui-router.github.io/ng1/docs/latest/interfaces/ng1.ng1statedeclaration.html#redirectto), or create your own Transition Hook ([such as `onEnter`](https://ui-router.github.io/ng1/docs/latest/interfaces/ng1.ng1statedeclaration.html#onenter)) which [returns a transition redirect](https://ui-router.github.io/ng1/docs/latest/modules/transition.html#hookresult).  You return a transition redirect by returning a [Target State](https://ui-router.github.io/ng1/docs/latest/classes/state.targetstate.html) from a transition hook.  You can [create a target state using `$state.target()`](https://ui-router.github.io/ng1/docs/latest/classes/state.stateservice.html#target)

```js
.state('foo', {
  url: '/foo?cachebust', 
  params: { cachebust: { dynamic: true } },

  redirectTo: function($transition$) {
    var injector = $transition$.injector();
    var $state = $transition$.stateService;
    // The redirectTo hook returns a promise for a target state.
    return injector.getAsync('MyResolve').then(function(resolvedata) {
      let params = angular.extend({}, $transition$.params(), { cachebust: Math.random() });
      if (resolvedata == 'Hey, Abbott!!!') {
        return $state.target($transition$.to(), params); // This is a redirect
      }
    });
  }
});
```

---
> 1.) In resolve or controller - Setting reloadOnSearch to false and update query string with $location.search().

This theoretically should have worked from the controller, as long as the transition was finished by the time the call to `.search()` was made.  `reloadOnSearch` sets all query parameters to `dynamic`.

---

> 2.) In resolve - Use $state.go with updated params object and notify:false option.


This will always cause the first transition to be superseded by the second because it's not completed yet.  This causes all the resolves to be re-fetched since it's a new transition, unrelated to the original one.

> `{locations: 'replace', reload: false, notify: false});`
None of these options do anything... 1) because it should be `location: 'replace'` 2) reload defaults to `false`, and `notify` is gone from 1.0
---

When we process redirects, if both transitions have common resolve data, the data is re-used instead of being re-fetched (See [this issue which describes the original problem](https://github.com/angular-ui/ui-router/issues/2274))

> I ended defining redirectTo function on state declaration. There I request my resolve to be executed and when promise is resolved, I redirect to the same state, with updated params (under some condition, not to cause a loop). 

That's perfect, nicely done.  I should have scanned down to your second reply before posting my suggestion, because it's almost *exactly the same* as what you ended up doing 👍 


>  But this still throws an error in console (visualizer shows state redirect):

I think that's a bug in the visualizer... would you mind checking?  Try turning off the visualizer and see if the warning still appears in the console.  If the warning disappears, can you file an issue with the visualizer project? https://github.com/ui-router/visualizer/issues

---

Closing because I think all questions are answered!

**Comment by SinisaG** [comment]
Thank you for detailed response @christopherthielen ! It gave me more insights on how to handle things, thanks. 

I've created a Plunker example, with the final solution (also one you recommend), to illustrate the promise rejection error (visualizer excluded). 

https://embed.plnkr.co/5Wb4IvyJWf9Gua6zKUqM/

The trace output, once `about` is clicked is 

```
1) Transition #0 r0: Started  -> "Transition#0( ''{} -> 'about.child'{"name":"chris"} )"

2) Transition #0 r0: <- Rejected "Transition#0( ''{} -> 'about.child'{"name":"chris"} )", reason: 
TransitionRejection(type: 2, message: The transition has been superseded by a different transition, detail: 'about.child'{"#":null,"name":"john","status":"OK"})

3) Transition #1 r0: Started  -> "Transition#1( ''{} -> 'about.child'{"name":"john","status":"OK"} )"

4) Transition #1 r0: <- Success  "Transition#1( ''{} -> 'about.child'{"name":"john","status":"OK"} )", final state: about.child
```

In Plunker, I don't get an error in console, although the trace shows the same message (`The transition has been superseded by a different transition`). Is this then correct? 

There is also no error with visualizer, so that's all good. :) I guess something else is logging this in console for us then, don't really know yet what it is. :) Thanks for your help!

**Comment by christopherthielen** [comment]
Yes, I see the error in the console now when I click "About" then immediately click "state with no resolve"
